src/tools/r2rdump/R2RHeader.cs

```diff
@@ -0,0 +1,126 @@
+        {
+            RelativeVirtualAddress = rva;
+            Offset = curOffset;
+
+            byte[] signature = new byte[sizeof(uint)];
+            GetField(pe, signature, curOffset, sizeof(uint));
+            SignatureString = System.Text.Encoding.Default.GetString(signature);
+            Signature = (uint)GetField(pe, ref curOffset, sizeof(uint));
+            if (Signature != READYTORUN_SIGNATURE)
+            {
+                throw new System.BadImageFormatException("Incorrect R2R header signature");
+            }
+
+            MajorVersion = (ushort)GetField(pe, ref curOffset, sizeof(ushort));
+            MinorVersion = (ushort)GetField(pe, ref curOffset, sizeof(ushort));
+            Flags = (uint)GetField(pe, ref curOffset, sizeof(uint));
+            NumberOfSections = (uint)GetField(pe, ref curOffset, sizeof(uint));
+
+            Sections = new R2RSection[NumberOfSections];
+            for (int i = 0; i < NumberOfSections; i++)
+            {
+                Sections[i] = new R2RSection((uint)GetField(pe, ref curOffset, sizeof(uint)),
+                    (uint)GetField(pe, ref curOffset, sizeof(uint)),
+                    (uint)GetField(pe, ref curOffset, sizeof(uint)));
+            }
+
+            Size = curOffset - Offset;
+        }
+
+        public override string ToString()
+        {
+            StringBuilder sb = new StringBuilder();
+            sb.AppendFormat($"Signature: 0x{Signature:X8} ({SignatureString})\n");
+            sb.AppendFormat($"RelativeVirtualAddress: 0x{RelativeVirtualAddress:X8}\n");
+            if (Signature == READYTORUN_SIGNATURE)
+            {
+                sb.AppendFormat($"Size: {Size} bytes\n");
+                sb.AppendFormat($"MajorVersion: 0x{MajorVersion:X4}\n");
+                sb.AppendFormat($"MinorVersion: 0x{MinorVersion:X4}\n");
+                sb.AppendFormat($"Flags: 0x{Flags:X8}\n");
+                sb.AppendFormat($"NumberOfSections: {NumberOfSections}\n");
+            }
+            return sb.ToString();
+        }
+
+        public long GetField(byte[] pe, int start, int size)
+        {
+            return GetField(pe, ref start, size);
+        }
+
+        public long GetField(byte[] pe, ref int start, int size)
+        {
+            byte[] bytes = new byte[size];
+            Array.Copy(pe, start, bytes, 0, size);
+            start += size;
+
+            if (size == 8)
+            {
+                return BitConverter.ToInt64(bytes, 0);
+            }
+            else if (size == 4)
+            {
+                return BitConverter.ToInt32(bytes, 0);
+            }
+            else if (size == 2)
+            {
+                return BitConverter.ToInt16(bytes, 0);
+            }
+            return bytes[0];
+        }
+
+        public void GetField(byte[] pe, byte[] bytes, int start, int size)
+        {
+            Array.Copy(pe, start, bytes, 0, size);
+        }
+    }
+
+    class R2RSection
```

Comment:
nit: This class would be a good candidate to be a struct instead of a class because it consists only of primitive types and is small. I would also support putting this in a separate file.

Reply:
`struct` / value types are another concept in the C# / CLR world that behave slightly different than reference types (classes) and using them effectively can improve performance of C# apps. They are best used in situations like this where the fields of the object are primitive types and small because the memory for structs are *not* allocated in the heap that the GC will have to keep track of, so doing this will mean one less thing the GC will have to track. Here is a small primer on the differences and of course the team can answer any questions you have, but this is one of those small C# things that are great to know about: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct